File: Best_time_to_buy_and_sell_stock.py
'''
Created on 2014-3-19


leetcoder : best time to buy and sell stock

Problem : Say you have an array for which the ith element is the price of a 
given stock on day i.

If you were only permitted to complete at most one transaction (ie, buy one and
 sell one share of the stock), design an algorithm to find the maximum profit.


@author: xqk
'''


class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        if len(prices) < 2:
            return 0;
        if len(prices) == 2 and prices[0]> prices[1]:
            return 0;
        maxprofit = 0;
        minprice = prices[0]
        for i in range(len(prices) - 1):
            if prices[i+1] < minprice:
                minprice = prices[ i + 1];
            else :
                profit = prices[i+1] - minprice;
                if profit > maxprofit:
                    maxprofit = profit
        return maxprofit
        
        
prices = [2, 5, 3, 8, 9, 4];
test = Solution();
print test.maxProfit(prices);

File: Cplusplus/Best_Time_to_Buy_and_Sell_Stock.cpp
/*
Leetcode : Best_Time_to_Buy_and_Sell_Stock.cpp

Say you have an array for which the ith element is the price

of a given stock on day i.

If you were only permitted to complete at most one transaction

(ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.
*/

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        if(prices.size() < 2)
            return 0;
        vector<int> cVec(prices.size() , 0);
        int minV = prices[0];
        cVec[0] = prices[0];
        for (int i = 1 ; i < prices.size() ; i++)
        {
            if(prices[i] < minV)
            {
                minV = prices[i];
            }
           cVec[i] = minV;
        }
        vector<int> eVec(prices.size(),0);
        int maxV = prices[prices.size()-1];
        eVec[eVec.size()-1] = prices[prices.size()-1];
        for(int i = prices.size() - 2 ; i >= 0 ; i--)
        {
            if(prices[i] > maxV)
            {
                maxV = prices[i];
               
            }
             eVec[i] = maxV;
        }
        int sum = 0;
        for(int i = 0 ; i < prices.size() - 1 ; i++)
        {
            if(eVec[i+1] - cVec[i] > sum)
                sum = eVec[i+1] - cVec[i];
        }
        return sum;
    }
};
